File: src/main/java/com/rtomyj/yugiohAPI/service/ProductService.java
package com.rtomyj.yugiohAPI.service;

import com.rtomyj.yugiohAPI.dao.database.Dao;
import com.rtomyj.yugiohAPI.helper.enumeration.products.ProductType;
import com.rtomyj.yugiohAPI.model.product.Product;
import com.rtomyj.yugiohAPI.model.product.Products;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;

@Service
public class ProductService
{

	@Autowired
	@Qualifier("jdbc")
	private Dao dao;


	public Products getProductsByLocale(final String locale)
	{
		final Products allProductsByLocale = new Products();
		allProductsByLocale.setProducts(dao.getProductsByLocale(locale));
		allProductsByLocale.setLinks();

		return allProductsByLocale;
	}

	public Products getAvailablePacks(final ProductType productType, final String locale)
	{
		final Products allPacks = dao.getAllProductsByType(productType, locale);
		allPacks.setLinks();

		return allPacks;
	}


	public Product getPack(final String packId, final String locale)
	{
		final Product pack = dao.getProductInfo(packId, locale);
		pack.getProductContent().addAll(dao.getProductContents(packId, locale));
		pack.setLinks();

		return pack;
	}

}

File: src/main/kotlin/com/rtomyj/skc/find/ProductService.kt
package com.rtomyj.skc.find

import com.rtomyj.skc.dao.ProductDao
import com.rtomyj.skc.model.MonsterAssociation
import com.rtomyj.skc.model.Product
import com.rtomyj.skc.model.ProductContent
import com.rtomyj.skc.skcsuggestionengine.TrafficService
import com.rtomyj.skc.util.enumeration.TrafficResourceType
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono

@Service
class ProductService @Autowired constructor(
  @param:Qualifier("product-jdbc") private val productDao: ProductDao,
  private val trafficService: TrafficService
) {
  fun getSingleProductUsingLocale(productId: String, locale: String, clientIP: String): Mono<Product> = Mono
      .zip(
        Mono.fromCallable { productDao.getProductInfo(productId, locale) },
        Mono.fromCallable { productDao.getProductContents(productId, locale) },
        trafficService.submitTrafficData(TrafficResourceType.PRODUCT, productId, clientIP)
      )
      .map {
        it.t1.productContent.addAll(it.t2)
        MonsterAssociation
            .transformMonsterLinkRating(
              it.t1
                  .productContent
                  .stream()
                  .filter { productContent -> productContent.card != null }
                  .map { productContent: ProductContent -> productContent.card!! }
                  .toList()
            )

        it.t1
      }
}